### DISCUSSION TITLE
How to indicate correct default values in mkdocstring

### DISCUSSION THREAD

**Discussion by ZhiyuanChen** [mkdocstrings/mkdocstrings#527] [Q&A]
Set default in arguments to mutable objects could cause [horrible errors](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/).
A general solution is to set default to None, and then assign correct default value in the function body. 
Although the correct default value should be properly addressed in docs to avoid ambiguation. 

However, it seems mkdocstring is not recognising the default values specified, like this:
<img width="849" alt="image" src="https://user-images.githubusercontent.com/28757366/215310152-217bd1e5-4b1f-4a23-bafb-6f722ac9c922.png">

**Comment by pawamoy** [comment]
Hello! Thanks, that's a good question.

Since mkdocstrings is meant to document APIs, in my opinion it makes sense that it says the default is `None` when it is specified as such in the signature. In my opinion again, telling that this `None` will be transformed into something else in the parameter description is the way to go.

If we were to somehow parse default values in the description, or in the function body, and display them instead of `None`, I believe that would be confusing to some users. They would wonder "what happens if I pass `None` instead? because that's what I get from the parent caller. should I pass it along or do I have to transform it myself first?", and they'll have to read the code to make sure.

Showing that the default is `None` makes it clear that if you don't pass anything, the function will handle it fine.

Just my opinion of course.

I'll convert this issue to a discussion.

**Comment by vinRosso** [comment]
I'm finding myself in the same situation and I also think that it would be more useful to parse the default values written in the docstrings.

Mkdocstrings renders *required* for mandatory attributes, so in my opinion it's natural to suppose that every other attribute could also be None. So in this case it's more useful in the documentation to see the default assumed value when the attribute isn't given.

```py
# example
def increment(var, amount=None):
    if not amount:
        amount = 2
    return var+amount
```
In this example, how am I supposed to know what the default increment is? I may suppose 1, which is wrong. Of course the default value would be written in the `Description` column, but it's not that immediate to see and understand rather then the `Default` column.

Obviously it would still be possible to leave it as `None` by "breaking" the parsing. Anyway allow both would be better imo.

**Reply by vinRosso** [reply]
Furthermore, the `None` is already visible in the signature, so i think it's redundant to have it also in the `Default` column.
![image](https://github.com/mkdocstrings/mkdocstrings/assets/22852084/68694884-5d15-4892-9300-5e8ba027d999)

**Reply by pawamoy** [reply]
Thanks for your feedback.

I don't think parsing the default value from the parameter description is feasible.

1. the language might not be english
2. even if we match some pattern, maybe parsing it is undesired (so this would require a config option, and fine-grain config is not yet well handled)
3. sometimes values you assign instead of `None`, in the function body, depend on the value of other parameters, so you write something like "defaults to X if A else Y if B else Z": how would you represent that in the default column? With text? So parsed defaults would always be raw text now, and not inline code?

I'm wondering if something based on [typing.Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated) would be better: `def f(a: Annotated[int | None, Default(2)] = None): ...` though that raises a lot of questions as well. Maybe `default` could be included in [@tiangolo's `typing.doc` spec](https://github.com/tiangolo/fastapi/blob/typing-doc/typing_doc.md#specification) somehow.

Another idea is to put the default value next to the parameter type, in the docstring, but then it means the type is duplicated in the signature and docstring:

```python
def f(a: int | None = None):
    """
    Args:
        a (int, default 2): Description.
    """
```

...because otherwise in `a (default 2)`, default 2 would be recognized as the type. But maybe we can improve the regex here.

**Reply by vinRosso** [reply]
I think the main point shouldn't be to cover every use case, but to give a basic feature that allows basic usage and customization.

The last prompt you gave is really nice and readable, I'm just wondering if it doesn't mess with the IDE's docstring identifiers used for auto-completion. But I'm not really getting why in the example you're using both type hinting and docstrings. Couldn't it be the same as doing the following? Sure regex could help here...
```py
def f(a = None):
    """
    Args:
        a (int | None, default 2): Description.
    """
```

Another possible solution could be to add a docstring section:
```py
def f(a = None):
    """
    Args:
        a (int | None): Description.

    Defaults:
        a: 2
    """
```
...this way you surely doesn't mess up with IDEs and it's maybe easier to parse.

**Reply by pawamoy** [reply]
I think having a Defaults section is a very good idea! It fits well within the docstring style standards and parsers :slightly_smiling_face: Could you open a feature request? At some point we'll need to document Griffe's parsers in details and reach out to the people who maintain the actual specs :upside_down_face: 

> But I'm not really getting why...

I meant that, when one uses annotations in the signature, they don't want to repeat them in the docstring. But writing the default without the type would be parsed as the type by the regex, not the default, so it kinda forces you to write the type as well, duplicated from the signature.
